Init system

rwsalie opened the pull request:
How `ck init <manifest name>` is designed:

Default args:
- description = (No description)
- suffix = json (over toml)
- kind = project

Double checking before the creation of the folder.

Rules:
- You can't `ck init` a component nor a target manifest outside of a project manifest.
- You can't overwrite the same manifest.
- `ck init` a component/target will be create at the most neighborhood manifest (in the worst scenario at the top of your project inside the src directories)

Cons:
- usage of `tomli_w` (a new package needed)
- usage of `__dict__` on Manifest (and his child) Class

rwsalie committed:
add `ck init` arguments

wip a smartter init system for cutekit

finish the smarter `ck init`, reduce the code size

Add description and correct schema

fix description

add toml support

Fix description and empty lines

lazyfix binary issue on json

:sparkles: `ck init` finish :sparkles:

fix prefer arg default value, remove useless loop

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+from . import cli, model, vt100, shell
+import dataclasses, os
```
When using `import` (without the `from`) you should do it one per line. 
Also the `dataclasses`  module isn't used

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+
+@cli.command("init", "Initialize your manifest.")
+def init_manifest(args: InitArgs):
+    if not vt100.ask(f"'{vt100.rgb(235, 111, 146) + args.name + vt100.RESET}' is the name of your project  ?"):
```
You should avoid using `+` when formatting
```suggestion
    if not vt100.ask(f"'{vt100.rgb(235, 111, 146)} {args.name}{vt100.RESET}' is the name of your project ?"):
```

When asking for a `--kind=exe` it will also be prompted (!= then a project)

Also is this prompt necessary ?

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+    if not vt100.ask(f"'{vt100.rgb(235, 111, 146) + args.name + vt100.RESET}' is the name of your project  ?"):
+        return 
+
+    filename = None
```
Missing type here

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+        return 
+
+    filename = None
+    manifest: Optional[Manifest] = None
```
```suggestion
    manifest: Optional[model.Manifest] = None
```

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+            )
+            setattr(manifest, "$schema", model.SUPPORTED_MANIFEST[1])
+
+    if hasattr(manifest, "description") and args.description:
```
I think there should be a more elegant way of doing that without doing a `hasattr`, since you are looking for `Component`, `Target` or `Project` (which all have the attribute). You could do perhaps a default case in your match.

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+
+    filename += f".{args.suffix}"
+    try:
+        if Path(os.getcwd(), filename).exists():
```
```suggestion
        if (Path.cwd() / filename).exists():
```

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+    TOML="toml"
+
+class InitArgs:
+    name: str = cli.operand("name", "Set the name of your project")
```
When the user doesn't provide any arguments and there is no default, the return type will be a `Optional[T]`

You also need to check if the option is not None otherwise the user will face a cryptic traceback saying that ck cannot perform operations with `NoneType`

keyboard-slayer commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+    name: str = cli.operand("name", "Set the name of your project")
+    description: str = cli.arg(None,"desc", "Set the description of your project", None)
+    kind: model.Kind = cli.arg(None, "kind", "Kind of the manifest", model.Kind.PROJECT)
+    suffix: Suffix = cli.arg(None, "prefer", "Manifest format to use", Suffix.JSON)
```
I don't really like the name suffix, maybe format would be better ?
`--format=json` `--format=toml`

sleepy-monax replied:
Also the name of the attribute should match the name of the argument

sleepy-monax commented on cutekit/project.py:
```diff
@@ -0,0 +1,106 @@
+            manifest = model.Project(
+                **manifest.__dict__
+            )
+            setattr(manifest, "$schema", model.SUPPORTED_MANIFEST[1])
```
Nah, don’t listen to that, just assign $schema after converting the manifest to a dict. Also, avoid using indexes on SUPPORTED_MANIFEST; instead, extract the constant, like model.PROJECT_SCHEMA.

rwsalie committed:
:hammer: fix few issues on project.py, add init test :hammer:

- add tests/test_init.py: a unit test for each init scenarios.
- fix project.py: never create other format when one is already used.
- fix project.py: id already used on a different manifest.
- remove: useless dataclasses import at project.py
- replace each Path(os.getcwd()) by Path.cwd()
- remove init directories when unable to write our manifest
- put every `ck init` at tests.

rwsalie committed:
:hammer: fix few issues on project.py, add init test :hammer:

- add tests/test_init.py: a unit test for each init scenarios.
- fix project.py: never create other format when one is already used.
- fix project.py: id already used on a different manifest.
- remove: useless dataclasses import at project.py
- replace each Path(os.getcwd()) by Path.cwd()
- remove init directories when unable to write our manifest
- put every `ck init` at tests.
- remove indirectional constant, useless shell.exec

sleepy-monax commented on cutekit/model.py:
```diff
@@ -37,7 +37,9 @@ class Kind(Enum):
-
+COMPONENT_SCHEMA = 0
+PROJECT_SCHEMA = 1
+TARGET_SCHEMA = 2
```
Let's avoid the indirection:

COMPONENT_SCHEMA = "https://schemas.cute.engineering/stable/cutekit.manifest.component.v1"
PROJECT_SCHEMA = "https://schemas.cute.engineering/stable/cutekit.manifest.project.v1"
TARGET_SCHEMA = "https://schemas.cute.engineering/stable/cutekit.manifest.target.v1"

SUPPORTED_MANIFEST = [
    COMPONENT_SCHEMA,
    PROJECT_SCHEMA,
    TARGET_SCHEMA
]

keyboard-slayer replied:
What do you think of using an enum instead ? (@sleepy-monax)

sleepy-monax replied:
Doesn't seem useful

rwsalie committed:
merge

rwsalie committed:
be compliant with mypy

- add `# types: ignore` for tomli_w, pytest's import.
- add `| None` for name, descriptions (they're nullable).
- put schema as PROJECT_SCHEMA, by default.
- remove redundant line `filename = "project"`

rwsalie committed:
fix runtime: remove nullable as attributes

rwsalie committed:
add ruff for project.py

sleepy-monax reviewed (approved):
LGTM, great work!
